### PR TITLE
add django example for python-socketio

### PR DESCRIPTION
**Pull request by sillygod**
I've resolved the conflict and make a pull request again. #24

### CODE REVIEW

**Comment by miguelgrinberg** [comment]
Oh, this is nice, awesome to have a Django example here. I'll look at this example in the next couple of days and let you know if I need you to make any changes. Thanks!

**Comment by sillygod** [comment]
@miguelgrinberg 

Thanks for this great implement socketio server with python. In fact, this example is my experimental project. I just find some trick or way to interact with django framework and I found a bug or problem in developing this example. 

That is, I can not send **message in Chinese** with `gevent` wsgi server but `eventlet` server works so in my example I use eventlet.

**Comment by miguelgrinberg** [comment]
Why did you write your example for Django 1.8 and not 1.9?

**Comment by sillygod** [comment]
@miguelgrinberg 

Q1:

There are many apis changed from version 1.8 to version 1.9 and not backward compatible more detail you see [here](https://docs.djangoproject.com/en/1.9/releases/1.9/). Actually, my company project still use django with version 1.8 because many third party library does not support django 1.9 although some top library may support (ex. rest_framework). That's why I choose django 1.8. For me, it is something like python 2 and python 3. lol.

Q2:

There is just small changed in my copy version `SocketMiddleware`. 

```
query = urllib.parse.parse_qs(environ.get('QUERY_STRING', ''))
sid = query.get('sid', None)

if sid is None:
    self.wsgi_app(environ, start_response)

```

I can not find any way to get framework specific instance (ex. django's request) so I make an entrance for that. see the following code snippet

```
@csrf_exempt
def socketio(request):
    try:
        request.environ['django_request'] = request
    except:
        logging.getLogger("socketio").error("Exception while handling socketio connection", exc_info=True)
    return HttpResponse(200)

urls = patterns("", (r'', socketio))
```

We must enter django's request-response cycle to get the `request` and set it to `environ`. 

```
query = urllib.parse.parse_qs(environ.get('QUERY_STRING', ''))
sid = query.get('sid', None)

if sid is None:
    self.wsgi_app(environ, start_response)

```
So first time, the request enter the middleware. it will not has sid in its query so let it enter the django request-response cycle.

To sum up,
I just find a trick to get framework specific instance.  Is there any way to get the same effect ? I would appreciate it if you tell me :)

Q3:

About that,  I will add it to the example. I forgot it.

Q4:

The reason I use class based style is I want to split socket related code and server code. In your example, you must define socket function in the file where server is defined. Furthermore, in my opinion, class based style is more easier to manage project's code. 

about this problem `The problem is that there are other decorators for error handling that don't appear to be supported in this model`. Sorry for my english. I can not fully understand your description.  Could you please give me a concrete example.

File: examples/django_chat/chat/management/commands/runserver_socketio.py [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+from time import sleep
+from os import getpid, kill, environ
+from signal import SIGINT
+import six
+import copy
+
+from django.conf import settings
+from django.core.handlers.wsgi import WSGIHandler
+from django.core.management.base import BaseCommand, CommandError
+from django.core.management.commands.runserver import naiveip_re, DEFAULT_PORT
```
Comment (miguelgrinberg) [inline examples/django_chat/chat/management/commands/runserver_socketio.py]: I believe the only issue when running your example under Django 1.9 is that `DEFAULT_PORT` isn't defined. At least that's what I see here when I test.

File: examples/django_chat/chat/sockets.py [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+
+@namespace('/test')
+class TestNamespace:
+
+    def __init__(self, name):
+        self.name = name
+        self.request = None  # django request object
+
+    def _get_socket(self, sid):
+        socket = namespace.server.eio._get_socket(sid)
```
Comment (miguelgrinberg) [inline examples/django_chat/chat/sockets.py]: The `eio._get_socket()` method is private. The `_` prefix indicates it is not meant to be called from the outside. Is there any way you can do what you need to do without using the internal socket object? I really don't like the idea of promoting its use in an example application.

File: examples/django_chat/chat/sockets.py [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+
+    def on_disconnect_request(self, sid):
+        namespace.server.disconnect(sid, namespace='/test')
+
+    # two method must have
+    def on_connect(self, sid, environ):
+        if 'django_request' in environ:
+            request = environ['django_request']
+            socket = self._get_socket(sid)
+            socket._request = request
```
Comment (miguelgrinberg) [inline examples/django_chat/chat/sockets.py]: It seems you are putting the Django request object as an extra member in the internal socket object. It is actually bad practice to inject new members into objects you get from another package. In addition to that, you are doing this on the connection event only, so in the case of long-polling, where new requests are issued all the time, this `_request` object that you saved would be stale.

File: examples/django_chat/sdjango/sd_middleware.py [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+
+        path = environ['PATH_INFO']
+        if path is not None and \
+                path.startswith('/{0}/'.format(self.engineio_path)):
+
+            query = urllib.parse.parse_qs(environ.get('QUERY_STRING', ''))
+            sid = query.get('sid', None)
+
+            if sid is None:
+                self.wsgi_app(environ, start_response)
```
Comment (miguelgrinberg) [inline examples/django_chat/sdjango/sd_middleware.py]: I think I understand what you are trying to do here, but I don't think this is a good solution. As I mentioned above, there isn't a single request, during the Socket.IO communication the client may send lots of requests, and here you are only saving the first. Not familiar with Django internals, but isn't there a way to create a request object from the `environ` dict?

File: examples/django_chat/sdjango/sd_manager.py [hunk 5]
```diff
@@ -0,0 +1,22 @@
+from socketio.base_manager import BaseManager
+
+
+class SdManager(BaseManager):
```
Comment (miguelgrinberg) [inline examples/django_chat/sdjango/sd_manager.py]: The `BaseManager` class is not supposed to be used from the outside. Besides, this subclass that you created doesn't do anything, so why do you have it?

File: examples/django_chat/chat/urls.py [hunk 6]
```diff
@@ -0,0 +1,13 @@
...
+    url, patterns, include
+)
+
+import sdjango
+
+from .views import socket_base
+
+
+urlpatterns = [
+    url(r'^socket\.io', include(sdjango.urls)),
```
Comment (miguelgrinberg) [inline examples/django_chat/chat/urls.py]: This defeats the purpose of using a middleware to do the routing between the socket and django endpoints. Note that in the middleware the `/socket.io` endpoint can be configured to be anything you want, while here it is hardcoded, so the ability to rename the endpoint is broken. I think I prefer a soluiton that does not require hacks around the middleware. I will think about how to get the request object using a cleaner way.

**Comment by miguelgrinberg** [comment]
> The reason I use class based style is I want to split socket related code and server code. In your example, you must define socket function in the file where server is defined.

You do not have to have the socket and http endpoints together, I have written larger applications that have them in separate modules or packages, while still using the decorator syntax for socket events.

>  Furthermore, in my opinion, class based style is more easier to manage project's code.

Well, yes, this is totally a matter of preference. The problem is that if you change the style to class-based, then a portion of this project's documentation is invalidated, and now you are asking people to follow a structure that is not documented.

File: examples/django_chat/sdjango/sd_manager.py (hunk 5 above)
Comment (sillygod) [inline examples/django_chat/sdjango/sd_manager.py]: yes, this can be removed. Originally, I plan to do some glue code for django here. However, it seems that nothing can do here.

File: examples/django_chat/sdjango/sd_middleware.py (hunk 4 above)
Comment (sillygod) [inline examples/django_chat/sdjango/sd_middleware.py]: In my understanding, it is true that the same client can send lots of requests but they are almost the same aren't they? That why I only save it once. 

I have no idea how to create request object from the environ.  I think it is no way. 
`request` instance is produced in django request-response cycle so we must enter wsgi app (django) to get it.

File: examples/django_chat/chat/sockets.py (hunk 3 above)
Comment (sillygod) [inline examples/django_chat/chat/sockets.py]: wow,  so in the case of long-polling, the `on_connect` function will be called multi times??

File: examples/django_chat/chat/sockets.py (hunk 2 above)
Comment (sillygod) [inline examples/django_chat/chat/sockets.py]: yes, I undertstand `_` prefix's meaning in python. I think every client has its own socket so I want the django request instance to bind to it

File: examples/django_chat/chat/sockets.py (hunk 3 above)
Comment (miguelgrinberg) [inline examples/django_chat/chat/sockets.py]: No, the connection event will happen only the first time. But in later requests, you will still be giving event handlers access to the connection request. It seems the correct thing to do would be either to provide access to the correct request object, or else not offer access to a request, since really this is a different path and the concept of an HTTP request makes really no sense.

**Comment by miguelgrinberg** [comment]
I have added a Django application that is a simplified version of this one, so I'm closing this PR.

**Comment by quietcoolwu** [comment]
@miguelgrinberg Hi Miguel, do we have django support now?(1.11 under python 2.7)

**Comment by miguelgrinberg** [comment]
@quietcoolwu Yes, and there is a Django example in this repository.
